Play again

td-extreme opened the pull request:
Added the ability for the user to play the game again after it finishes.

td-extreme committed:
Added new__game method to GameBoard class.

td-extreme committed:
Added the option to play again.

Added play_again? to TicTacToe class.
Added logic to play_game to check if the player wants to play again:

mikeknep commented on lib/tictactoe.rb:
```diff
@@ -20,6 +20,16 @@ def initialize (game_io, player1, player2)
+    @io.puts_message("Play again (Y/N) : ")
+    input = @io.get_input
+    return true if (input == "y" || input == "Y")
+    false
```
You have this `return true if ...  \n false` pattern in a few places. Since equality checks and logical operators both return booleans in Ruby, you can simplify this to just `input == "y" || input == "Y"`. There are a few other places where you could take advantage of OR's behavior to "cascade" through possible return values.

dadarek replied:
Ditto Mike. Also - may be overkill here - but you can take advantage of the ENumerable module if you were to have more choices here (i.e. `["y", "Y"].include?(input)`)

td-extreme replied:
I need to stop myself from doing that. I feel like I keep doing this with boolean returns. I changed the code to use the ENumerable module.

mikeknep commented on spec/tictactoe_spec.rb:
```diff
@@ -7,7 +7,22 @@
+    test_ttt.play_game
+  end
+
+  describe "the play_again? method " do
```
I'd prefer `play_again?` be private since no other class depends on it. This would mean that the behavior needs to be tested higher up inside the game loop. Were you able to figure out the RSpec multiple calls feature? It'll be necessary here to prevent an infinite loop...

dadarek replied:
Ditto Mr. Mike again. You're already injecting the thing that decides whether or not to play again (the `game_io`) - so you should easily be able to stub that to return "Y" the first time and "N" the second time, and then somehow assert that the game was played twice.

Example:
`allow(something).to receive(:some_method).and_return( 1, 2, 3 )`
will return 1 the first time `some_method` is called, 2 the second time, and then 3 for all subsequent calls to `some_method`

Note that this is different from:
`allow(something).to receive(:some_method).and_return( [1, 2, 3] )`
which will always return the array `[1, 2, 3]` no matter how many times it's called.

td-extreme replied:
I changed `play_again?` to private and made the `new_game` method now is part of the `TicTacToe` class. It is also private. Both are tested through the `play_game` method of `TicTacToe`

dadarek commented on lib/tictactoe.rb:
```diff
@@ -20,6 +20,16 @@ def initialize (game_io, player1, player2)
   def play_game
+    begin
     @game_loop.game_loop
+    @board.new_game
```
Instead of mutating the board, would it make sense to throw away and spin up a new one?

Generally, if you're "starting again" or "repeating a process", things get simpler to understand and debug if you just dump everything you don't need anymore and start again. In the long term, this prevents lots of headaches.

td-extreme replied:
I changed the code so that it creates a new `GameBoard` object. I had to add `attr_accessor :board` to both the `GameLoop` and `GameScreen` classes as they need a reference to the `GameBoard` object. So when `TicTacToe` creates a new `GameBoard` it can update the reference in those other to classes to point to the current board.

dadarek commented:
@td-extreme - it's a good sign that, in order to add this feature, you only had to touch a handful of files. Anytime you're adding features or fixing bugs, the number of files you're touching is representative of the cohesion and coupling of your classes.

If you actually stop mutating the board (get rid of `new_game`) and just create a fresh one for each new game, then you don't even have to touch the `Board` class altogether.

Nice job!

(Let's see how many files you have to touch for the next feature 😄 )

Please address some of the comments before merging. Wait for a 👍  from @mikeknep  and I.

td-extreme committed:
Removed the new_game method from GameBoard.

td-extreme committed:
Added new_game to TicTacToe. Helpers made private.

Made the helper methods of play_again? and new_game private.
Testing for those occurs through the play_game method.

mikeknep commented on lib/tictactoe.rb:
```diff
@@ -11,15 +11,32 @@ def initialize (game_io, player1, player2)
+    begin
     @game_loop.game_loop
+    new_game
+    end while (play_again?)
```
Nitpick: you don't need the parens here.

mikeknep commented on spec/tictactoe_spec.rb:
```diff
@@ -7,7 +7,23 @@
+  it "plays the game twice " do
+    expect(test_ttt.game_loop).to receive(:game_loop).twice
+    allow(test_ttt).to receive(:new_game)
+    expect(test_ttt).to receive(:play_again?).and_return(true, false)
```
Nice job here. However, I think you should stub the `@io.get_input` call instead of `play_again?`. The reason is if you stub `play_again?`, you aren't testing that method's functionality anywhere—it's stubbed out in every test so the actual implementation (returning `true` with some "yes" response) is never called/tested. By stubbing `get_input` instead, you're basically saying OK, I have some other object (`GameIO`) with its own behavior and that behavior is under test separately. Assuming I get a certain input from that object and method, this class under test here (`TTT`) will behave like so: (insert expectation).

Alternatively to all of the above, you could instantiate your `TicTacToe` object with a custom `MockIO`.

(This applies to all three tests here.)

td-extreme replied:
I switched it to stub the GameIO instead.

td-extreme committed:
Changed test to stub game_io instead of play_again

dadarek commented on lib/tictactoe.rb:
```diff
@@ -11,15 +11,32 @@ def initialize (game_io, player1, player2)
   def play_game
+    begin
     @game_loop.game_loop
+    new_game
```
fix spacing

dadarek commented on lib/tictactoe.rb:
```diff
@@ -11,15 +11,32 @@ def initialize (game_io, player1, player2)
+  def play_again?
+    @io.puts_message("Play again (Y/N) : ")
+    input = @io.get_input
+    ["Y", "y", "yes", "YES", "Yes"].include?(input)
```
nice!

dadarek commented on spec/tictactoe_spec.rb:
```diff
@@ -7,7 +7,25 @@
+  end
+
+  it "plays the game twice " do
+    expect(test_ttt.game_loop).to receive(:game_loop).twice
```
A little awkward to stub a method (or variable) of the thing you're testing.

Don't worry about this for now, but just keep that in mind for the future - it's sometimes a sign that you're doing something funky somewhere.

td-extreme replied:
I didn't change the name of the test. It had a game_loop in my `TicTacToe` class before, but I made a separate `GameLoop` class. The method I'm testing in `TicTacToe` is the `play_move` method, which does make a call to the `GameLoop` class. I should have changed the code to `it "test play_game" do`

td-extreme replied:
I guess it is slightly confusing, as there is a GameLoop class that is a loop that plays the game. And the TicTacToe has a loop that calls that loop and then does a play again loop. So there are two separate game loops.

dadarek commented on spec/tictactoe_spec.rb:
```diff
@@ -7,7 +7,25 @@
+
+  it "gameboad is blank after new_game " do
+    expect(test_ttt.game_loop).to receive(:game_loop).twice
+    expect(test_ttt).to receive(:play_again?).and_return(true, false)
```
Again, stubbing variables/methods on the thing under test is usually a smell. If you're injecting the `game_io` and that's the thing that ultimately controls whether or not a game is played again, then you would probably just stub that.

mikeknep replied:
I think this may be a little unclear, so just to clarify: when you `expect(something).to receive(:some_method)`, RSpec is actually stubbing that method under the hood—that's how it manages to count how many times the method is called. For example:

``` ruby
class Foo
  def method_1
    2.times { method_2 }
    true
  end

  private

  def method_2
    puts "hello world"
  end
end


describe Foo do
  let(:foo) { Foo.new }

  it "this will print 'hello world' twice to the console when running specs" do
    expect(foo.method_1).to be true
  end

  it "this will NOT print 'hello world' to the console because the method is stubbed by RSpec" do
    expect(foo).to receive(:method_2).twice
    foo.method_1
  end
end
```

dadarek commented:
@td-extreme nice job.

👍 from me. @mikeknep is going to be out tomorrow, Thursday, and Friday - so if he doesn't respond by tomorrow morning then go ahead and merge.

mikeknep commented:
I added a note about RSpec stubbing, but it's not a blocker for me. Ship it!

As @dadarek mentioned, I'll be out the rest of the week, but keep up the great work—I think the evolution of this PR today has been great.

td-extreme committed:
Updated test name.
